### ISSUE TITLE
Commando build stacking

### ISSUE THREAD

**Issue by MeSaber** [beyond-all-reason/Beyond-All-Reason#1909]
Giving a build command will stack air units and also build 1 per each Commando selected.

Selecting 2 Commando to build a transport will end in 2 stacked transports, with no visual confirmation that its 2 of em.

I dont think its suppose to work like this? I would rather see Commandos build air units the same as ground units. I guess this is a part of the air-stacking 'bug' where air units dont have collision(?) on ground so they end up in the same place.

**Comment by KyleAnthonyShepherd** [comment]
Yup, side effect of how aircraft collision and footprint flags are set to prevent getting stuck on air factories. 
Although, still not perfect. There is an air-vs-air collision flag that the engine uses, and can be set in unitdefs, but CANNOT be set and changed through any existing lua gadget function.

**Comment by efrec** [comment]
I think I understand the issue but not the potential solutions.

- We are not likely to use decollision of air units outside this one case, so I'd rather not ask the engine to support toggling it on/off via lua, but that is an option.
- We can destack build orders pretty easily, actually; in `AllowCommand`, check the map position, query units in a tiny radius around it (`CallAsTeam(unitTeam, Spring.GetUnitsInRadius, x, z, Game.squareSize, -3)`), and check if any of them match the building def id. This is how the engine checks whether a build order was already placed and changes the order to a `CMD_REPAIR`/build assist instead.
- We can make landed, stacked air units nudge away from each other, or at least nudge away from a unit being built on top of them.

Several options, each fairly easy to approach. After some discussion, I'd like to mark this as a good first issue.

**Comment by MeSaber** [comment]
May or may not be related: This is twitchers behaving very odd to queued build orders of mobile units, doing the same thing, building too many of a selected unit, not planes this time though.

[2025-08-04_02-35-01-008_Avalanche 3_2025.04.08.sdfz.zip](https://github.com/user-attachments/files/21569789/2025-08-04_02-35-01-008_Avalanche.3_2025.04.08.sdfz.zip)

My guess is that each twitcher makes its own 'list' of units it wants to build in the same spots, lingering to some extent beyond a unit finishing, its most prone to error when you do 2x3 apparently, but not so much in 2x2. You can still see the ghost images of wanting to build further units though but is cancelled in 2x2.

**Comment by efrec** [comment]
Seems relevant to me.

That also seems solvable by converting build orders into repair orders for the same def id. It would work for unstacking mines when they are set not to collide, too, so there are a few benefits to that change.

The pattern of canceled build commands is interesting, too, but I have no guesses for the cause.

**Comment by sprunk** [comment]
> I'd rather not ask the engine to support toggling it on/off via lua, but that is an option.

Sounds useful to have in engine regardless and it may involve just flipping a bool internally.

**Comment by MeSaber** [comment]
@efrec well example: a constructor is built and it queues 3 mexes towards frontline, the con starts building the first one and commander passes it building say 3rd mex in same place. At some point the con has to drop the 3rd mex order as its already there, either by knowing a mex been built or basically unplaceable due to already a building there.

Maybe the cancel time is just too long, in twitcher case the grunt gotta move first before its possible to build there again which is the ghost image lingering meanwhile grunt moves, however even if grunt moved it still decides to not build there.

The last part in the replay i uploaded shows twitcher in a row getting to build queue in different times and by that maybe get different timings on the cancel time. What happened is 2x2 then ended in like 2-3 extra grunts. Why dunno :)

In commando case there is no moving of planes due to no collision, ending up in exactly as many planes as commandos.

**Comment by KyleAnthonyShepherd** [comment]
> > I'd rather not ask the engine to support toggling it on/off via lua, but that is an option.
> 
> Sounds useful to have in engine regardless and it may involve just flipping a bool internally.

Agree. I believe the bool set by unitdef `collide` cannot be set or changed by Lua, while Lua can already set and change a bunch of other blocking and collision tags via `SetUnitBlocking`

**Comment by efrec** [comment]
> > > I'd rather not ask the engine to support toggling it on/off via lua, but that is an option.
> > 
> > 
> > Sounds useful to have in engine regardless and it may involve just flipping a bool internally.
> 
> Agree. I believe the bool set by unitdef `collide` cannot be set or changed by Lua, while Lua can already set and change a bunch of other blocking and collision tags via `SetUnitBlocking`

I thought there was another move type flag, besides `collide`, that dealt with air units trying to avoid collisions when pathing rather than bumping together in huddles like ground units do (edit: I may have been thinking of detection for `lastCollidee` is all). Either way, both would apply.

**Comment by efrec** [comment]
Marking this as a good first issue. This is a nice bundle of related issues that are solvable with an isolated lua gadget. I'd consider the engine concerns as separate, for now.
